include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp

```diff
@@ -980,6 +980,45 @@ SpatialOperatorBase<dim, Number>::prescribe_initial_conditions(VectorType & velo
   pressure = pressure_double;
 }
 
+template<int dim, typename Number>
+void
+SpatialOperatorBase<dim, Number>::interpolate_analytical_solution(VectorType & velocity,
+                                                                  VectorType & pressure,
+                                                                  double const time) const
+{
+  AssertThrow(field_functions->analytical_solution_velocity,
+              dealii::ExcMessage("FieldFunctions::analytical_solution_velocity not set"));
+  AssertThrow(field_functions->analytical_solution_pressure,
+              dealii::ExcMessage("FieldFunctions::analytical_solution_pressure not set"));
+
+  velocity.zero_out_ghost_values();
+  pressure.zero_out_ghost_values();
+
+  field_functions->analytical_solution_velocity->set_time(time);
+  field_functions->analytical_solution_pressure->set_time(time);
+
+  // This is necessary if Number == float
+  using VectorTypeDouble = dealii::LinearAlgebra::distributed::Vector<double>;
+
+  VectorTypeDouble velocity_double;
+  VectorTypeDouble pressure_double;
+  velocity_double = velocity;
+  pressure_double = pressure;
+
+  dealii::VectorTools::interpolate(*get_mapping(),
+                                   get_dof_handler_u(),
+                                   *(field_functions->analytical_solution_velocity),
+                                   velocity_double);
+
+  dealii::VectorTools::interpolate(*get_mapping(),
+                                   get_dof_handler_p(),
+                                   *(field_functions->analytical_solution_pressure),
+                                   pressure_double);
+
+  velocity = velocity_double;
+  pressure = pressure_double;
```

Comment:
This code is more or less the same as above for the function `prescribe_initial_conditions()`. Hence, I would suggest to write one private member function (that takes the velocity and pressure functions as arguments and that implements these lines of code), which is then called by the public member functions `prescribe_initial_conditions()` and `interpolate_analytical_solution()`.

Reply:
Instead let's finally introduce a free function to remove the boilerplate code and avoid some unnecessary copy operations in numerous places of ExaDG. I would like to do this in a follow up PR if it is OK for you :)